### DISCUSSION TITLE
Approve/refuse new user via email using one time URIs

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#16391] [Q&A] [answered]
Hi,

I'm writing APIs for a headless CMS project using Orchard Core 1.8.3. I have used the `UserManager` (Microsoft.AspNetCore.Identity) to provide basic functions that allow the Admin to manage user roles via APIs.

Now, I'm adding a new feature: when a new user registers on the website, an email is sent to the Admin with two URIs: one to approve (enable) the user, and the other to refuse (leave the user disabled, or delete it. We still decide what to do).
**What I'm looking for is a way to make those URIs usable only once, and only for that specific purpose.**

1. Would it be a good idea to use the `GenerateUserTokenAsync` and `VerifyUserTokenAsync` method (here is the [GenerateUserTokenAsync](https://github.com/aspnet/AspNetIdentity/blob/4eed61d83eb960a1805f5aa2933d44671a8706a9/src/Microsoft.AspNet.Identity.Core/UserManager.cs#L1356), and here the [VerifyUserTokenAsync](https://github.com/aspnet/AspNetIdentity/blob/4eed61d83eb960a1805f5aa2933d44671a8706a9/src/Microsoft.AspNet.Identity.Core/UserManager.cs#L1333) implementation) to achieve the one-time-use URIs behavior described above?

2. From what I have seen in the Orchard Core `RegistrationController` ([here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore.Modules/OrchardCore.Users/Controllers/RegistrationController.cs#L101)), **Orchard Core doesn't provide the ability to approve or refuse a new user via email**. If the `Users Registration` feature is turned on, and the setting "_Users must be approved before they can log in_" is checked, the new user is registered as disabled, and the approval (enabling) is done by the Admin directly from the backoffice. There's no email that the Admin receives to perform the approve/refuse action, right?

**Any suggestions or best practices for implementing this?**

Thank you so much

**Comment by Piedone** [comment]
1. I'd start with these too. However, I have no experience building such a feature so I'm not sure.
2. Yep, there's no such feature built-in.

That being said, if it's already admins who approve/refuse users, then I don't really see the point of adding a token to the URL (unless you want only the recipients of the e-mail to be able to manage these users, not any admin with the necessary permissions, what I'd instead recommend). Instead, I'd only include GET links in the e-mail, pointing to an admin UI where they have the two operations as two buttons (or just link to the given user's built-in edit screen on the OC admin).

Otherwise, you're binding a state-changing operation to a GET request, what should otherwise be a POST or something else.

**Reply by MarGraz** [reply]
@Piedone thank you for your reply.

**My client requirements** is that the bookstore Admin can approve/refuse new users by email. I want to ensure that the URIs in the email are not reusable multiple times, and are specific for the single new user. If I provide a simple endpoint with the UserId as a parameter, it can be used with other users, multiple times, which I want to avoid.

**My environment:** currently I have several microservices based on Orchard Core 1.8.3, and I am working on two of them: the "SSO" microservice and the "Users" microservice, which manages different kinds of users for the bookstore (but doesn't store the user it self).

I have created two API endpoints in my SSO microservice: one to create the token, and one to validate it. These endpoints are used by the other microservice (like Users), which calls the SSO via POST method to get the token, and is authenticated using `OpenIddict`.

Here are the two endpoints:

```csharp
[Route("api")]
[Authorize(AuthenticationSchemes = "Api"), IgnoreAntiforgeryToken]
[ApiController]
public class UserTokenApiController : Controller
{
    private readonly ITokenHandler _tokenHandler;

    public UserTokenApiController(ITokenHandler tokenHandler)
    {
        _tokenHandler = tokenHandler;
    }

    [HttpPost]
    [Route("tokenGenerate")]
    public async Task<IActionResult> GenerateTokenAsync([FromForm] string userId, [FromForm] string purpose)
    {
        return Ok(await _tokenHandler.GenerateTokenAsync(userId, purpose));
    }

    [HttpPost]
    [Route("tokenVerify")]
    public async Task<IActionResult> VerifyTokenAsync([FromForm] string userId, [FromForm] string purpose, [FromForm] string token)
    {
        return Ok(await _tokenHandler.VerifyTokenAsync(userId, purpose, token));
    }
}
```

Here is the implementation in the handler:

```csharp
 public async Task<string> GenerateTokenAsync(string userId, string purpose)
 {
     IUser? user = await _userManager.FindByIdAsync(userId);
     if (user == null)
     {
         throw new InvalidOperationException("User not found.");
     }

     return await _userManager.GenerateUserTokenAsync(user, "Default", purpose);
 }

 public async Task<bool> VerifyTokenAsync(string userId, string purpose, string token)
 {
     IUser? user = await _userManager.FindByIdAsync(userId);
     if (user == null)
     {
         throw new InvalidOperationException("User not found.");
     }

     return await _userManager.VerifyUserTokenAsync(user, "Default", purpose, token);
 }
```

**The problem I'm facing now** is that a "Default" token provider doesn't exist in Orchard Core. 
I get this exception:

```csharp
System.NotSupportedException: No IUserTwoFactorTokenProvider<TUser> named 'Default' is registered.
   at Microsoft.AspNetCore.Identity.UserManager`1.GenerateUserTokenAsync(TUser user, String tokenProvider, String purpose)
```

I also tried to instantiate a `DataProtectorTokenProvider` in the `Program.cs`, but I get another exception, and Orchard Core didn't start. This is because Orchard Core already implement and configure his own Identity:

```csharp
InvalidOperationException: Scheme already exists: Identity.Application
Microsoft.AspNetCore.Authentication.AuthenticationOptions.AddScheme(string name, Action<AuthenticationSchemeBuilder> configureBuilder)
```

**So the question is**: 

- What's the correct way to get the Orchard Core Identity to instantiate a new custom Token Provider? 🤔

Thank you

**Reply by Piedone** [reply]
I don't know this from the top of my head, sorry, but I would suggest you consult the code of `UserManager`, as well as look up how such token providers are normally used in an ASP.NET Core app. While there can be curiosities, as you've discovered, by and large, it shoudlw work the same in OC.

**Reply by MarGraz** [reply]
@Piedone I have the same problem that @MikeAlhayek had in [this issue](https://github.com/OrchardCMS/OrchardCore/issues/7717). 

**Basically, I'm not able to add a `DataProtectorTokenProvider`**. When I look into the `UserManager` after the app starts, the only two available token providers added by Orchard Core, are `Email` and `Phone`, as shown in the screenshot below. **But what if I want to add another token provider?**

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/cb3289c5-8f0e-4a28-8b9d-15aa741278b0)

I don't know how to add a `DataProtectorTokenProvider`, because **every normal procedure I try to apply in the `Program.cs` that usually works in .NET, returns errors in OC**. For example, doing this:

```csharp
services.AddIdentity<User, Role>()
            .AddTokenProvider<DataProtectorTokenProvider<User>>(TokenOptions.DefaultProvider); // TokenOptions.DefaultProvider is a string that has value "Default"
```

**Return an exception** that involve `YesSql`: 

> **InvalidOperationException: Cannot consume scoped service 'Microsoft.AspNetCore.Identity.ILookupNormalizer' from singleton 'YesSql.Indexes.IIndexProvider'.**
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitScopeCache(ServiceCallSite scopedCallSite, CallSiteValidatorState state)

**Questions:**

- Could this exception be a bug in YesSql? 🤔
- What if I pass `Email` as token provider in my two handler "GenerateToken" and "VerifyToken" methods, [here](https://github.com/OrchardCMS/OrchardCore/discussions/16391#discussioncomment-9941632)?

Thank you

**Reply by Piedone** [reply]
I don't have a ready answer for this, sorry. Maybe others will be able to help.

**Reply by MikeAlhayek** [reply]
@MarGraz if you upgrade to 2.0.0-previews, we register a `Default` provider by default. https://github.com/OrchardCMS/OrchardCore/blob/7b4120ed52156d503bde24ffcf60d1bc59597b11/src/OrchardCore.Modules/OrchardCore.Users/Startup.cs#L168

Alternatively, you register a default provider yourself. However, in 2.0.0-preview there are improvements around the token providers which can improve security and allow you to configure them. more on that can found in the [2.0.0 release notes,](https://docs.orchardcore.net/en/latest/releases/2.0.0/#users-module_2)

**Reply by MarGraz** [reply, marked as answer]
@Piedone @MikeAlhayek @weirdyang sorry for the late reply, I totally forgot to update this discussion 😅

@MikeAlhayek, unfortunately, I can't use the 2.0.0-previews because I'm going into production soon. I'm not comfortable using an unstable release for production.

In the end, **I created my own token handler** and passed it in Dependency Injection as a singleton. Of course, **if the server reboots, I will lose the token list**, but this is not a problem for my requirements since the tokens are not meant to be permanent.

I used the `Purpose` value to be able to generate multiple tokens for the same `UserId`. This could be useful if you need it in multiple parts of your code.

Here is the **interface I made** for my custom token service:

```csharp
public interface ITokenService
{
    /// <summary>
    /// Generates a token for a specified user and purpose.
    /// If a token already exists for the specified user and purpose, the existing token is returned.
    /// Otherwise, a new token is generated. Token older that 3 days will be automatically deleted.
    /// </summary>
    /// <param name="userId">The ID of the user for whom the token is being generated.</param>
    /// <param name="purpose">The purpose for which the token is being generated.</param>
    /// <returns>The generated token.</returns>
    string GenerateToken(string userId, string purpose);

    /// <summary>
    /// Validates a token for a specified user and purpose.
    /// The token is considered valid if it matches the specified user and purpose and is not expired.
    /// Once validated, the token is removed from the storage to ensure it is only used once.
    /// </summary>
    /// <param name="token">The token to validate.</param>
    /// <param name="userId">The ID of the user for whom the token was generated.</param>
    /// <param name="purpose">The purpose for which the token was generated.</param>
    /// <returns><c>true</c> if the token is valid; otherwise, <c>false</c>.</returns>
    bool ValidateToken(string token, string userId, string purpose);
}
```

This is the **implementation of my custom TokenService**:

```csharp
  public class OneTimeTokenService : ITokenService
 {
     private class TokenInfo
     {
         public string Token { get; set; }
         public string UserId { get; set; }
         public string Purpose { get; set; }
         public DateTime CreationDate { get; set; }
     }

     // ConcurrentDictionary is used to prevent multiple instances and ensure that the token service is truly a singleton. It is useful for thread-safe operations and make sure the service is registered as a singleton in the DI container
     private readonly ConcurrentDictionary<string, TokenInfo> _tokens = new ConcurrentDictionary<string, TokenInfo>();
     private DateTime _lastCleanupDate = DateTime.MinValue;

     public string GenerateToken(string userId, string purpose)
     {
         CleanupOldTokens();

         // If for a specific userId and purpose a token already exists, it will return it 
         TokenInfo? existingToken = _tokens.Values.FirstOrDefault(t => t.UserId == userId && t.Purpose == purpose);
         if (existingToken != null)
         {
             return existingToken.Token;
         }

         string token = Guid.NewGuid().ToString();
         TokenInfo tokenInfo = new TokenInfo
         {
             Token = token,
             UserId = userId,
             Purpose = purpose,
             CreationDate = DateTime.UtcNow
         };

         _tokens[token] = tokenInfo;
         return token;
     }

     public bool ValidateToken(string token, string userId, string purpose)
     {
         if (_tokens.TryGetValue(token, out var tokenInfo))
         {
             if (tokenInfo.UserId == userId && tokenInfo.Purpose == purpose)
             {
                 _tokens.TryRemove(token, out _);
                 return true;
             }
         }
         return false;
     }

     /// <summary>
     /// Cleans up tokens that are older than 3 days.
     /// This method only executes once per day.
     /// </summary>
     private void CleanupOldTokens()
     {
         DateTime currentDate = DateTime.UtcNow.Date;
         
         if (_lastCleanupDate < currentDate)
         {
             DateTime cutoffDate = currentDate.AddDays(-3);
             List<string> oldTokens = _tokens.Where(t => t.Value.CreationDate < cutoffDate).Select(t => t.Key).ToList();

             foreach (var oldToken in oldTokens)
             {
                 _tokens.TryRemove(oldToken, out _);
             }

             _lastCleanupDate = currentDate;
         }
     }
 }
```

This is **how I register it** in the `Program.cs` to pass it in DI:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services
    .AddOrchardCms()
 .ConfigureServices( services =>
 {
    // Add custom services
     services.AddSingleton<ITokenService, OneTimeTokenService>();
     
     [...]
 }
```

Thank you so much 😜

**Reply by Piedone** [reply]
Thanks for sharing!

**Comment by weirdyang** [comment]
a possible alternative might be to use `ISecurityTokenService` to generate a time limited token with a custom payload https://github.com/OrchardCMS/OrchardCore/blob/52dca204ee223fb8495dcd1e3fae03a4b1129fbd/src/OrchardCore.Modules/OrchardCore.Workflows/Services/SecurityTokenService.cs#L8
